_posts/2018-10-15-12-2-Use-SkipToken-for-Server-Side-Paging.md

```diff
@@ -0,0 +1,94 @@
+
+`GET ~/Products/$skip=70&$top=10`
+#### Server-driven paging
+In server-driven paging, the client asks for a collection of entities and the server sends back partial results as well as a nextlink to use to retrieve more results. The nextlink is an opaque link which may use $skiptoken to identify the last loaded record.
+### Problem
+Currently, WebAPI uses $skip for server-driven paging which is a slight deviation from the OData standard and can be problematic when the data source can get updated concurrently. For instance, a deletion of a record may cause the last record to be sent down to the client twice. 
+### Proposed Solution
+WebAPI will now implement $skiptoken. When a collection of entity is requested which requires paging, we will assign the key value of the last sent entity to $skiptoken in the nextlink url prepended by values of the orderby properties in same order. While processing a request with $skiptoken, we will add another condition to the predicate based on the value of the skipoken. 
+### Technical details
+After all the query options have been applied, we determine if the results need pagination. If the results need pagination, we will pass the generated skiptoken value based off of the last result to the method that generates the nextpage link.   
+
+#### Format of the nextlink
+The nextlink may contain $skiptoken if the result needs to be paginated. In WebAPI the $skiptoken value will be a list of pairs, where the pair consists of a property name and property value separated by a delimiter(:). The orderby property and value pairs will be followed by key property and value pairs in the value for $skiptoken. Each property and value pair will be comma separated.
+```
+~/Products?$skiptoken=Id:27
+~/Books?$skiptoken=ISBN:978-2-121-87758-1,CopyNumber:11
+~/Products?$skiptoken=Id:25&$top=40
+~/Products?$orderby=Name&$skiptoken=Name:'KitKat',Id:25&$top=40
+~/Cars(id)/Colors?$skip=4
+```
+We will not use $skiptoken if the requested resource is not an entity type. Rather, normal skip will be used. 
+
+This is the default format but services can define their own format for the $skiptoken as well but in that case, they will have to parse and generate the skiptoken value themselves.
+
+#### Generating the nextlink
+The next link generation method in ___GetNextPageHelper___ static class will take in the $skiptoken value along with other query parameters and generate the link by doing special handling for $skip, $skiptoken and $top. It will pass on the other query options as they were in the original request.
+
+##### 1. Handle $skip
+We will omit the $skip value if the service is configured to support $skiptoken and a collection of entity is being requested. This is because the first response would have applied the $skip query option to the results already. 
+##### 2. Handle $top
+We will reduce the value of $top query option by the page size if it is greater than the page size.   
```

Comment:
If the $top query option is smaller than the page size, we should not return a nextlink.